lib/class-wp-rest-menu-locations-controller.php

```diff
@@ -0,0 +1,269 @@
+<?php
+/**
+ * REST API: WP_REST_Menu_Locations_Controller class
+ *
+ * @subpackage REST_API
+ * @package    WordPress
+ */
+
+/**
+ * Core class used to access menu locations via the REST API.
+ *
+ * @see   WP_REST_Controller
+ */
+class WP_REST_Menu_Locations_Controller extends WP_REST_Controller {
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->namespace = 'wp/v2';
+		$this->rest_base = 'menu-locations';
+	}
+
+	/**
+	 * Registers the routes for the objects of the controller.
+	 *
+	 * @see   register_rest_route()
+	 */
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base,
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_items' ),
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+					'args'                => $this->get_collection_params(),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/(?P<location>[\w-]+)',
+			array(
+				'args'   => array(
+					'location' => array(
+						'description' => __( 'An alphanumeric identifier for the menu location.' ),
+						'type'        => 'string',
+					),
+				),
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_item' ),
+					'permission_callback' => array( $this, 'get_item_permissions_check' ),
+					'args'                => array(
+						'context' => $this->get_context_param( array( 'default' => 'view' ) ),
+					),
+				),
+				'schema' => array( $this, 'get_public_item_schema' ),
+			)
+		);
+	}
+
+	/**
+	 * Checks whether a given request has permission to read menu locations.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|bool True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) {
+		if ( 'edit' === $request['context'] ) {
+			$types = get_post_types( array( 'show_in_rest' => true ), 'objects' );
```

Comment:
Could you explain the reasoning behind using this permissions check?

Reply:
As opposed to something like `edit_theme_options`.

Reply:
I am going to loop in @schlessera to explain this one

Reply:
The WP capabilities are a mystery to me, especially when you look at what the REST API is doing in some places, and it has been a while since I worked on that code. I guess I either a) thought this makes the most sense based on what options I considered, or (more likely) b) this was what was included in the original code I copy-pasted...

Sorry I can't tell you much more than that. The code I provided was a quick first draft I cobbled together from existing controllers. So if you see something that does not make immediate sense, it might never have made sense at all to begin with! ;)

Reply:
It looks like the permission checks used in the types controller. I think for now it'd make sense to use `edit_theme_options` and reevaluate later if we need to loosen that check.